gensim/models/phrases.py

```diff
@@ -130,54 +236,68 @@ def __str__(self):
-                            (sentence_no, total_words, len(vocab)))
             sentence = [utils.any2utf8(w) for w in sentence]
-            for bigram in zip(sentence, sentence[1:]):
-                vocab[bigram[0]] += 1
-                vocab[delimiter.join(bigram)] += 1
-                total_words += 1
+            if sentence_no % 10000 == 0:
+                # TODO in parallel version, log information must be changed
+                logger.info("PROGRESS: at sentence #%i" % (sentence_no + start_position))
 
-            if sentence:  # add last word skipped by previous loop
-                word = sentence[-1]
-                vocab[word] += 1
+            for token_bigram in zip(sentence, sentence[1:]):
+                sub_vocab[token_bigram[0]] += 1
+                sub_vocab[delimiter.join(token_bigram)] += 1
+                total_words += 1
 
-            if len(vocab) > max_vocab_size:
-                utils.prune_vocab(vocab, min_reduce)
-                min_reduce += 1
+            if sentence:  # add last sentence token skipped by previous loop
+                last_token = sentence[-1]
+                sub_vocab[last_token] += 1
+                total_words += 1
 
-        logger.info("collected %i word types from a corpus of %i words (unigram + bigrams) and %i sentences" %
-                    (len(vocab), total_words, sentence_no + 1))
-        return min_reduce, vocab
+        # if iterator is empty, let parent method know by raising StopIteration.
+        if sentence_no == -1:
+            raise StopIteration()
+        logger.info(
+            ("processed chunk of sentences contains %i new word types " +
+             "(unigrams + bigrams) from a corpus of %i tokens and %i sentences") % (
+                len(sub_vocab), total_words, sentence_no + 1))
+        # return built token=>freq dictionary        
+        return sub_vocab
 
     def add_vocab(self, sentences):
         """
         Merge the collected counts `vocab` into this phrase detector.
 
         """
-        # uses a separate vocab to collect the token counts from `sentences`.
-        # this consumes more RAM than merging new sentences into `self.vocab`
-        # directly, but gives the new sentences a fighting chance to collect
-        # sufficient counts, before being pruned out by the (large) accummulated
-        # counts collected in previous learn_vocab runs.
-        min_reduce, vocab = self.learn_vocab(sentences, self.max_vocab_size, self.delimiter)
-
-        logger.info("merging %i counts into %s", len(vocab), self)
-        self.min_reduce = max(self.min_reduce, min_reduce)
-        for word, count in iteritems(vocab):
-            self.vocab[word] += count
-        if len(self.vocab) > self.max_vocab_size:
-            utils.prune_vocab(self.vocab, self.min_reduce)
-            self.min_reduce += 1
+        sentences_iterator = iter(sentences)
+        step = int(self.max_vocab_size / self.average_sentence_size)
```

Comment:
The logic for using step-sizes that are `max_vocab_size` (previously a count of unique tokens) divided by `average_sentence_size` (which isn't really known without a full-scan, and includes non-unique token counts) isn't clear. If `step` is actually an important parameter for performance or memory use, maybe allow it to be specified directly?